File: src/infrastructure/memory/memoryWalletRepository.ts
import type { Id, Wallet } from '../../domain/entities/types'
import type { WalletRepository } from '../../application/ports/repositories'
import type { Snapshotable } from './memoryRepositories'

const clone = <T>(value: T): T => structuredClone(value)

/** محافظ في الذاكرة — للاختبار بلا فايربيز (CLAUDE.md #6). */
export class MemoryWalletRepository
  implements WalletRepository, Snapshotable<Wallet[]>
{
  private items = new Map<Id, Wallet>()

  constructor(seed: readonly Wallet[] = []) {
    for (const w of seed) this.items.set(w.id, clone(w))
  }

  async listAll(): Promise<Wallet[]> {
    return [...this.items.values()].map(clone)
  }

  async findById(id: Id): Promise<Wallet | null> {
    const found = this.items.get(id)
    return found ? clone(found) : null
  }

  async save(wallet: Wallet): Promise<void> {
    this.items.set(wallet.id, clone(wallet))
  }

  snapshot(): Wallet[] {
    return [...this.items.values()].map(clone)
  }

  restore(state: Wallet[]): void {
    this.items = new Map(state.map((w) => [w.id, clone(w)]))
  }
}

File: native-app/app/src/commonMain/kotlin/app/masroufy/memory/MemoryWalletRepository.kt
package app.masroufy.memory

import app.masroufy.core.Id
import app.masroufy.core.Wallet
import app.masroufy.port.WalletRepository

/** محافظ في الذاكرة — نقل `memoryWalletRepository.ts`. */
class MemoryWalletRepository(seed: List<Wallet> = emptyList()) : WalletRepository {
    private val items = LinkedHashMap<Id, Wallet>()

    init {
        for (w in seed) items[w.id] = w
    }

    override suspend fun listAll(): List<Wallet> = items.values.toList()

    override suspend fun findById(id: Id): Wallet? = items[id]

    override suspend fun save(wallet: Wallet) {
        items[wallet.id] = wallet
    }
}
